### PR TITLE
8282567: Improve source-date handling in build system

### PR DESCRIPTION
**Pull request by magicus**
When doing reproducible builds, controlling the build time is imperative. To make this as easy as possible, some changes are needed in the build system.

 * If source-date is set to anything other than "updated" (meaning that it should be determined at build time), then the default value of --with-hotspot-build-time should be the same time.

* If the industry standard SOURCE_DATE_EPOCH is set when running configure, the default value for --with-source-date should be picked up from this variable. (If the command line option is given, it will override the environment variable however.)

* Finally the code can be made a bit clearer that we can set and export SOURCE_DATE_EPOCH to SOURCE_DATE in spec.gmk, unless we're using the "updated" (determined at build time) value.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8282567](https://bugs.openjdk.java.net/browse/JDK-8282567): Improve source-date handling in build system

### Reviewers
 * [Erik Joelsson](https://openjdk.java.net/census#erikj) (@erikj79 - **Reviewer**)
 * [Severin Gehwolf](https://openjdk.java.net/census#sgehwolf) (@jerboaa - **Reviewer**) ⚠️ Review applies to b97625fd5f3466160b979ec5a7f01609fd306db4

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/7660/head:pull/7660` \
`$ git checkout pull/7660`

Update a local copy of the PR: \
`$ git checkout pull/7660` \
`$ git pull https://git.openjdk.java.net/jdk pull/7660/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 7660`

View PR using the GUI difftool: \
`$ git pr show -t 7660`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/7660.diff">https://git.openjdk.java.net/jdk/pull/7660.diff</a>

</details>

### CODE REVIEW

**Comment by magicus** [comment]
To clarify, the end effect of these changes is that building OpenJDK will basically be compliant with the method of just setting SOURCE_DATE_EPOCH. (The caveat is that it must be set at configure time, not build time.) So
```
$ export SOURCE_DATE_EPOCH=123
$ bash configure
$ make
```
will cause the build to default to building in reproducible mode, with the date given by SOURCE_DATE_EPOCH.

**Review by erikj79 — APPROVED** [review]

**Comment by jerboaa** [comment]
> To clarify, the end effect of these changes is that building OpenJDK will basically be compliant with the method of just setting SOURCE_DATE_EPOCH. (The caveat is that it must be set at configure time, not build time.) So
> 
> ```
> $ export SOURCE_DATE_EPOCH=123
> $ bash configure
> $ make
> ```

Do you have an example configure output when used like that post-patch?

> will cause the build to default to building in reproducible mode, with the date given by SOURCE_DATE_EPOCH.

Hmm, when building via RPM, `SOURCE_DATE_EPOCH` is usually set. If I understand this correctly, prior to this, the RPM build would be non-reproducible. After it it would be. That may be confusing to some. If post-patch this makes the build reproducible, does it mention that via a WARNING/INFO or anything like that after this patch? I don't see it. Am I missing it? It would be good to have something to that effect in the configure output.

**Comment by erikj79** [comment]
I think building.md needs to be updated to reflect these changes, as you no longer need to manually propagate the value of SOURCE_DATE_EPOCH through --with-source-date as the documentation currently suggests.

**Comment by magicus** [comment]
@jerboaa 

The output in configure can look like this:
```
checking what source date to use... 946684800, from SOURCE_DATE_EPOCH
checking for --enable-reproducible-build... enabled, default
```

But your comment made me think. I believe this should be more prominently displayed. My latest update therefore adds a new line to the summary. So now it looks like this:

```
* Version string: 19-internal-adhoc (19-internal)
* Source date:    946684800 (2000-01-01T00:00:00Z)
```

If no fixed source date is set, this is shown instead:
```
* Version string: 19-internal-adhoc (19-internal)
* Source date:    Determined at build time
```

(This is still the default if --with-source-date is not given or SOURCE_DATE_EPOCH is not set.)

**Comment by magicus** [comment]
As for RPM builds (by whom?) suddenly being build with --enable-reproducible-build, I would not think it is a matter of concern. For linux, the only real difference --enable-reproducible-build does is add the --date flag to jmod (and to the bootjdk jar, if it supports it), and runs a special tool on src.zip to make it reproducible. The only real reason this flag exists (and not just is the default behavior) is out of an abundance of caution. On Windows, it triggers the `-experimental:deterministic` flag, which has the scary sounding word "experimental" in it. But we have been running with this option enabled for all Oracle Windows builds for years by now, I think, without seeing any problems.

So I'm actually contemplating removing the --enable-reproducible-build flag, and just make that behavior default. There's very little reason not to have this turned on always, and many good reasons for having it.

**Comment by magicus** [comment]
@erikj79 You are absolutely correct that the documentation needed updating. Thank you for reminding me. When I did that, I also realized I've missed the --with-copyright-year flag, so this is now also set automatically from the source-date. With this change, now we can *really* build reproducible just by setting --with-source-date or SOURCE_DATE_EPOCH.

**Comment by jerboaa** [comment]
> As for RPM builds (by whom?)

We, Red Hat, build OpenJDK in Fedora and RHEL via `rpmbuild` :)

> suddenly being build with --enable-reproducible-build, I would not think it is a matter of concern. For linux, the only real difference --enable-reproducible-build does is add the --date flag to jmod (and to the bootjdk jar, if it supports it), and runs a special tool on src.zip to make it reproducible. The only real reason this flag exists (and not just is the default behavior) is out of an abundance of caution. On Windows, it triggers the `-experimental:deterministic` flag, which has the scary sounding word "experimental" in it. But we have been running with this option enabled for all Oracle Windows builds for years by now, I think, without seeing any problems.
> 

Fair enough.

> So I'm actually contemplating removing the --enable-reproducible-build flag, and just make that behavior default. There's very little reason not to have this turned on always, and many good reasons for having it.

+1. It would be more explicit than by sneaking this change in by starting to respect `SOURCE_DATE_EPOCH` (which it didn't before - at least consistently)

**Review by jerboaa — APPROVED** [review]
Looks fine to me.

File: doc/building.md [hunk 1]
```diff
@@ -1527,57 +1527,85 @@ https://reproducible-builds.org) for more information about the background and
...
-lacking.
+To generate reproducible builds, you must set `SOURCE_DATE_EPOCH` before running
+`configure`. The value in `SOURCE_DATE_EPOCH` will be stored in the
+configuration, and used by `make`. Setting `SOURCE_DATE_EPOCH` before running
+`make` will have no effect on the build.
 
-  * Do not rely on `configure`'s default adhoc version strings
+You must also make sure your build does not rely on `configure`'s default adhoc
+version strings. Default adhoc version strings `OPT` segment include user name,
+source directory and timestamp. You can either override just the `OPT` segment
```
Comment (erikj79) [inline doc/building.md:1546]: The timestamp comes from Oracle internal builds using Jib. If you just use configure, the default OPT string is 'adhoc.<username>.<base dir name>'.

As a followup on this, I think we should change jib-profiles.js to extract the date from the buildId and set --with-source-date based on it.

File: doc/building.md [hunk 2]
```diff
@@ -1527,57 +1527,85 @@ https://reproducible-builds.org) for more information about the background and
...
-`configure`. The default value is `updated`, which means that
-`SOURCE_DATE_EPOCH` will be set to the current time each time you are running
-`make`.
+ * `--with-source-date`
 
-The [`SOURCE_DATE_EPOCH` environment variable](
-https://reproducible-builds.org/docs/source-date-epoch/) is an industry
-standard, that many tools, such as gcc, recognize, and use in place of the
-current time when generating output.
+    This option controls how the JDK build sets SOURCE_DATE_ EPOCH when
```
Comment (erikj79) [inline doc/building.md:1570]: Space in variable name. Should it be backtick quoted?

**Review by erikj79 — APPROVED** [review]

**Pull request closed without merging**